### PR TITLE
[Clang][NEON] Add neon target guard to intrinsics

### PR DESCRIPTION
**Pull request by Lukacma**
This patch improves reported error when NEON intrinsics are used without neon target feature.

### CODE REVIEW

File: clang/test/Sema/aarch64-neon-without-target-feature.cpp [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  vbfdot_f32(v2f32, v4bf16, v4bf16); // expected-error {{always_inline function 'vbfdot_f32' requires target feature 'neon'}}
+  vcreate_bf16(10);
+  vdup_lane_bf16(v4bf16, 2); // expected-error {{'__builtin_neon_splat_lane_bf16' needs target feature bf16,neon}}
+  vdup_n_bf16(bf16); // expected-error {{always_inline function 'vdup_n_bf16' requires target feature 'neon'}}
+  vld1_bf16(0); // expected-error {{'__builtin_neon_vld1_bf16' needs target feature bf16,neon}}
+  vcvt_f32_bf16(v4bf16); // expected-error {{always_inline function 'vcvt_f32_bf16' requires target feature 'neon'}}
+  vcvt_bf16_f32(v4f32); // expected-error {{always_inline function 'vcvt_bf16_f32' requires target feature 'neon'}}
+  vmull_p64(poly64, poly64);  // expected-error {{always_inline function 'vmull_p64' requires target feature 'neon'}}
+  vmull_high_p64(poly64x2, poly64x2);  // expected-error {{always_inline function 'vmull_high_p64' requires target feature 'neon'}}
+  vtrn1_s8(v8i8, v8i8); // expected-error {{always_inline function 'vtrn1_s8' requires target feature 'neon'}}
```
Comment (CarolineConcatto) [inline clang/test/Sema/aarch64-neon-without-target-feature.cpp:31]: Can you add tests for the fails reported here:
https://github.com/llvm/llvm-project/issues/98152#issuecomment-2218280104
https://godbolt.org/z/Y4Eebeqz7 - vbslq
https://godbolt.org/z/3zWdcsMKK - vqabsq
https://godbolt.org/z/3zWdcsMKK
https://godbolt.org/z/Y4Eebeqz7

**Review by CarolineConcatto — COMMENTED** [review]
Thank you Marian for the patch.
I left some comments.

File: clang/test/Sema/aarch64-neon-without-target-feature.cpp [hunk 2]
```diff
@@ -0,0 +1,33 @@
+// RUN: %clang_cc1 -triple aarch64-none-linux-gnu -target-feature +dotprod  -target-feature +fullfp16 -target-feature +fp16fml -target-feature +i8mm -target-feature +bf16 -verify -emit-llvm -o - %s
+
+// This test is testing the diagnostics that Clang emits when compiling without '+neon'.
```
Comment (mgabka) [inline clang/test/Sema/aarch64-neon-without-target-feature.cpp:3]: do we have a positive Sema tests to show that the intrinsics are available when the neon feature is present?

Comment (Lukacma) [reply clang/test/Sema/aarch64-neon-without-target-feature.cpp:3]: No. But I think this is already covered by CodeGen tests for individual intrinsics, which need to have +neon to work.

File: clang/include/clang/Basic/arm_neon.td [hunk 3]
```diff
@@ -289,7 +289,7 @@ def SPLATQ : WInst<"splat_laneq", ".(!Q)I",
                    "UcUsUicsilPcPsfQUcQUsQUiQcQsQiQPcQPsQflUlQlQUlhdQhQdPlQPl"> {
   let isLaneQ = 1;
 }
-let TargetGuard = "bf16" in {
+let TargetGuard = "bf16,neon" in {
```
[the remark this answers is not in the document]
Comment (Lukacma) [reply clang/include/clang/Basic/arm_neon.td:292]: Yes that's right.

File: clang/test/Sema/aarch64-neon-without-target-feature.cpp (hunk 1 above)
Comment (Lukacma) [reply clang/test/Sema/aarch64-neon-without-target-feature.cpp:31]: Done

**Review by CarolineConcatto — APPROVED** [review]

**Pull request merged**